zebra/if_netlink.c

```diff
@@ -458,6 +464,91 @@ uint32_t kernel_get_speed(struct interface *ifp, int *error)
 	return get_iflink_speed(ifp, error);
 }
 
+static ssize_t
+netlink_gre_set_msg_encoder(struct zebra_dplane_ctx *ctx, void *buf,
+			    size_t buflen)
+{
+	struct {
+		struct nlmsghdr n;
+		struct ifinfomsg ifi;
+		char buf[];
+	} *req = buf;
+	uint32_t link_idx;
+	unsigned int mtu;
+	struct rtattr *rta_info, *rta_data;
+	struct interface *gre_ifp;
+	struct zebra_if *zif;
+	struct zebra_l2info_gre *gre_info;
+
+	if (buflen < sizeof(*req))
+		return 0;
+	memset(req, 0, sizeof(*req));
+
+	req->n.nlmsg_type =  RTM_NEWLINK;
+	req->n.nlmsg_len = NLMSG_LENGTH(sizeof(struct ifinfomsg));
+	req->n.nlmsg_flags = NLM_F_REQUEST;
+
+	req->ifi.ifi_index = dplane_ctx_get_ifindex(ctx);
+
+	gre_ifp = if_lookup_by_index(req->ifi.ifi_index,
```

Comment:
I don't think this is safe: this function will be called in the dplane pthread, right? this api is not thread-safe. if you need info from these kinds of datastructs, that are owned and maintained in the context of the zebra main pthread, you have to take a snapshot of the info when constructing the dplane ctx, since that occurs in the main pthread.

Reply:
ok, I stored gre l2 info in dplane context.